SYCL/ESIMD/api/functional/common.hpp

```diff
@@ -0,0 +1,92 @@
+//      tests -------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// Common file for tests on simd class.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/intel/experimental/esimd.hpp>
+#include <sycl/sycl.hpp>
+
+#include "../../esimd_test_utils.hpp"
+#include "logger.hpp"
+#include "type_coverage.hpp"
+#include "type_traits.hpp"
+#include "value.hpp"
+
+#include <vector>
+
+namespace esimd_test {
+namespace api {
+namespace functional {
+
+// Dummy kernel for submitting some code into device side.
+template <typename DataT, int NumElems, typename T> struct Kernel;
+
+template <typename DataT>
+using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+
+template <typename DataT>
+using shared_vector = std::vector<DataT, shared_allocator<DataT>>;
+
+// Calls simd constructor in provided invocation context, which depends on the
+// TestCaseT type. TestCaseT is a struct, that should have call_simd_ctor method
+// that should return constructed object of simd class.
+// This function returns std::vector instance with the output data.
+template <typename DataT, int NumElems, typename TestCaseT>
+auto call_simd(sycl::queue &queue, const shared_vector<DataT> &ref_data) {
+
+  shared_vector<DataT> result{NumElems, shared_allocator<DataT>{queue}};
+
+  queue.submit([&](sycl::handler &cgh) {
+    const auto ref = ref_data.data();
+    auto out = result.data();
+
+    cgh.single_task<Kernel<DataT, NumElems, TestCaseT>>(
+        [=]() SYCL_ESIMD_KERNEL {
+          sycl::ext::intel::experimental::esimd::simd<DataT, NumElems>
+              result_simd =
+                  TestCaseT::template call_simd_ctor<DataT, NumElems>(ref);
+          result_simd.copy_to(out);
+        });
+  });
+  return result;
+}
+
+// Bitwise comparing two provided variables.
+template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
+  constexpr size_t size{sizeof(T)};
+
+  const auto &lhs_bytes = reinterpret_cast<const unsigned char(&)[size]>(lhs);
+  const auto &rhs_bytes = reinterpret_cast<const unsigned char(&)[size]>(rhs);
+
+  bool result{true};
+  for (size_t i = 0; i < size; ++i) {
+    result &= lhs_bytes[i] == rhs_bytes[i];
+  }
+  return result;
+}
+
+// Function that comparing two provided variables. If this variable is floating
+// point type or signed type this variables will be bitwise compared, othervise
+// variables will be compared by "==" operator.
+template <typename T> bool are_equal(T lhs, T rhs) {
+  if constexpr (type_traits::is_floating_point_v<T> || std::is_signed_v<T>) {
+    return are_bitwise_equal(lhs, rhs);
```

Comment:
```suggestion
// A wrapper to speed-up bitwise comparison
template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
  if constexpr (type_traits::is_floating_point_v<T> || std::is_signed_v<T>) {
    return details::are_bitwise_equal(lhs, rhs);
```
or may be we don't need a separate wrapper at all?

General comment - please avoid duplicating code logic in comments.

Reply:
I tell You about this wrapper, in the test plan **all** types should be bitwise compared 
What do You think about this function? should we delete it and bitwise comparing all data types or this implementation is ok?
The comment for this function will be updated in one of next commits

Reply:
this comment was applied in this commit: 848b25756a3ca77feefbb306b2b620e384439b46
also use update can be found here: 1f1f8b9d296773e2495ab9330c86dad1f180df93